File: src/app/services/developpeurService.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Developpeur } from '../models/developpeur';

@Injectable({
  providedIn: 'root'
})
export class DeveloppeurService {
  private apiUrl = 'http://localhost:8091/api/developpeurs'; 
  constructor(private http: HttpClient) {}

  // 🔹 Récupérer tous les développeurs
  getAll(): Observable<Developpeur[]> {
    return this.http.get<Developpeur[]>(this.apiUrl);
  }

  // 🔹 Récupérer un développeur par ID
  getById(id: number): Observable<Developpeur> {
    return this.http.get<Developpeur>(`${this.apiUrl}/${id}`);
  }

  // 🔹 Créer un développeur
  create(dev: Developpeur): Observable<Developpeur> {
    return this.http.post<Developpeur>(this.apiUrl, dev);
  }

  // 🔹 Modifier un développeur
  update(id: number, dev: Developpeur): Observable<Developpeur> {
    return this.http.put<Developpeur>(`${this.apiUrl}/${id}`, dev);
  }

  // 🔹 Supprimer un développeur
  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // 🔹 Rechercher par nom
  searchByName(keyword: string): Observable<Developpeur[]> {
    const params = new HttpParams().set('keyword', keyword);
    return this.http.get<Developpeur[]>(`${this.apiUrl}/search`, { params });
  }

  // 🔹 Recherche globale
  searchGlobal(keyword: string): Observable<Developpeur[]> {
    const params = new HttpParams().set('keyword', keyword);
    return this.http.get<Developpeur[]>(`${this.apiUrl}/filter`, { params });
  }
searchByNom(nom: string): Observable<Developpeur[]> {
  return this.http.get<Developpeur[]>(`${this.apiUrl}/search?nom=${nom}`);
}

  // 🔹 Filtrer par spécialité
  findBySpecialite(specialite: string): Observable<Developpeur[]> {
    return this.http.get<Developpeur[]>(`${this.apiUrl}/specialite/${specialite}`);
  }

  // 🔹 Récupérer les développeurs par projet
  findByProjet(projetId: number): Observable<Developpeur[]> {
    return this.http.get<Developpeur[]>(`${this.apiUrl}/projet/${projetId}`);
  }
}

File: src/main/java/sen/saloum/promeet/services/Impl/DeveloppeurService.java
package sen.saloum.promeet.services.Impl;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import sen.saloum.promeet.dto.DeveloppeurDTO;
import sen.saloum.promeet.mapstracts.DeveloppeurMapper;
import sen.saloum.promeet.models.Developpeur;
import sen.saloum.promeet.repos.DeveloppeurRepository;
import sen.saloum.promeet.repos.ProjetRepository;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Transactional
public class DeveloppeurService {

    private final DeveloppeurRepository developpeurRepository;
    private final ProjetRepository projetRepository;
    private final DeveloppeurMapper developpeurMapper;

    public DeveloppeurService(DeveloppeurRepository developpeurRepository, 
                              ProjetRepository projetRepository,
                              DeveloppeurMapper developpeurMapper) {
        this.developpeurRepository = developpeurRepository;
        this.projetRepository = projetRepository;
        this.developpeurMapper = developpeurMapper;
    }

    // 🔹 Créer un développeur
    public DeveloppeurDTO create(DeveloppeurDTO dto) {
        Developpeur dev = developpeurMapper.toEntity(dto);
        dev = developpeurRepository.save(dev);
        return developpeurMapper.toDto(dev);
    }

    // 🔹 Mettre à jour un développeur
    public DeveloppeurDTO update(Long id, DeveloppeurDTO dto) {
        Developpeur existing = developpeurRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Développeur non trouvé"));
        existing.setNom(dto.getNom());
        existing.setEmail(dto.getEmail());
        existing.setNiveau(dto.getNiveau());
        existing.setSpecialite(dto.getSpecialite());
        existing.setActif(dto.isActif());
        Developpeur updated = developpeurRepository.save(existing);
        return developpeurMapper.toDto(updated);
    }

    // 🔹 Supprimer un développeur
    public void delete(Long id) {
        developpeurRepository.deleteById(id);
    }

    // 🔹 Récupérer tous les développeurs
    public List<DeveloppeurDTO> getAll() {
        return developpeurRepository.findAll()
                .stream()
                .map(developpeurMapper::toDto)
                .collect(Collectors.toList());
    }

    // 🔹 Récupérer un développeur par ID
    public Optional<DeveloppeurDTO> getById(Long id) {
        return developpeurRepository.findById(id).map(developpeurMapper::toDto);
    }

    // 🔹 Recherche par nom
    public List<DeveloppeurDTO> searchByName(String keyword) {
        return developpeurRepository.searchByName(keyword)
                .stream()
                .map(developpeurMapper::toDto)
                .collect(Collectors.toList());
    }

    // 🔹 Recherche globale
    public List<DeveloppeurDTO> searchGlobal(String keyword) {
        return developpeurRepository.searchGlobal(keyword)
                .stream()
                .map(developpeurMapper::toDto)
                .collect(Collectors.toList());
    }

    // 🔹 Récupérer par spécialité
    public List<DeveloppeurDTO> findBySpecialite(String specialite) {
        return developpeurRepository.findBySpecialiteIgnoreCase(specialite)
                .stream()
                .map(developpeurMapper::toDto)
                .collect(Collectors.toList());
    }


}
